### DISCUSSION TITLE
In a Lua filter, how to access previous (or next) element?

### DISCUSSION THREAD

**Discussion by pagiraud** [jgm/pandoc#9867] [Q&A] [answered]
I’m writing a Lua filter, where I use the Figure function to change the figure appearance if it has a given class. But I also would like to change appearance of the previous element in the AST if it’s a header. How can I achieve that? And, reversely, if I select a header if it has a certain class, can I modify the next element in the AST if it’s a figure?

Excerpt of my AST:

```json
, Header 2 ( "section" , [ "plain" ] , [] ) []
, Figure
    ( "" , [] , [] )
    (Caption
       Nothing
       [ Plain
           [ Str "Le"
           , Space
           , Str "petit"
           , Space
           , Str "\226ne"
           , Space
           , Str "gris"
           ]
       ])
    [ Plain
        [ Image
            ( "" , [ "fullframegraphicnoadjust" ] , [] )
            [ Str "Le"
            , Space
            , Str "petit"
            , Space
            , Str "\226ne"
            , Space
            , Str "gris"
            ]
            ( "images/somepicture.png" , "" )
        ]
    ]
```

Thanks for your time.

**Comment by kysko** [comment]
In this precise case, it is possible, if you hierarchicalize the doc using [`make_sections`](https://pandoc.org/lua-filters.html#pandoc.structure.make_sections), which creates sections Divs out of Headers.

In the first problem, let's say the target Figure class is `foo`. You then ask the question:
> Is there a section Div (so a Div with class `section`) for which the **second Block** (in its content) is a Figure with class `foo`?

If so, then modify the **first** Block of the Div, which by construction must be the Header associated with the section.

In the second problem, let's say the target Header class is `bar`. You then ask the question:
> Is there a section Div so that the **first Block** has class `bar`, and which has a Figure as **second Block**?

If so, then modify that second Block.

After this, you can undo the `make_sections` if you must.

Of course, by "ask the question" above, I mean a well-crafted filter.

~/Edit: hmm... i see you already have class `section` for the header, but this is not why I mentioned class `section` for section Div above, which is created by `make_sections`.~
/Edit 2: oops again... misread the Native format: 'section' is your Header identifier. That shouldn't matter.

**Reply by pagiraud** [reply]
Thanks for your reply. It may be a solution. Actually, this may be a bit overkill but at least I realized that there are two different `make_sections` : `pandoc.utils.make_sections` (that I knew of) and `pandoc.structure.make_sections` (that I didn’t know).

**Comment by massifrg** [comment, marked as answer]
You may search that combination of Header+Figure with [Blocks](https://pandoc.org/lua-filters.html#filters-on-element-sequences).

**Reply by pagiraud** [reply]
Thanks a lot. That is exactly what I needed. I had already read that and the example about [removing spaces before citations](https://pandoc.org/lua-filters.html#remove-spaces-before-citations) but didn’t see I was trying to achieve something similar.

This is the code I used to test this solution:

```lua
local figuretype
local function is_figure_in_given_frame(frame, figure)
  if frame and frame.t == 'Header'
    and figure and (figure.t == 'Figure' or figure.t == 'Para')
    and frame.classes[1] == "given" then
    local figtype
    figtype = figure.t
    return figtype
  end
end

function Blocks (blocks)
  -- Go from end to start to avoid problems with shifting indices.
  for i = #blocks-1, 1, -1 do
    figuretype = is_figure_in_given_frame(blocks[i], blocks[i+1])
    if figuretype == "Figure" then --if there is a caption
      print(blocks[i+1].t)
    elseif figuretype == "Para" then --if there is no caption
      print(blocks[i+1].t)
    end
  end
  return blocks
end
```
Thanks to you I finished my filter!
